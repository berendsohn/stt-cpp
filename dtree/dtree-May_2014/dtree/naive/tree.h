// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_NAIVE_TREE_H_
#define DTREE_NAIVE_TREE_H_

#include "dtree/naive/common.h"

namespace naive {

template<typename Base>
class EndTree : public Base {
 public:
  EndTree() : parent_(NULL) {}
  EndTree* parent() const { return parent_; }
  void set_parent(EndTree* parent) { parent_ = parent; }
  template<typename Forwarder>
  void CopyFrom(const EndTree& node, Forwarder forward);
 private:
  EndTree* parent_;
  DISALLOW_COPY_AND_ASSIGN(EndTree);
};

template<typename Type, typename Base>
class WithStaticValue : public Base {
 public:
  typedef dtree::Nop<Type> Group_;
  Type value() const { return value_; }
  void set_value(Type value) { value_ = value; }
  //
  template<typename Node>
  static Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, Type value);
 protected:
  WithStaticValue() : value_() {}
 private:
  Type value_;
};

template<typename Group, typename Base>
class WithAncValue : public Base {
 public:
  typedef Group Group_;
  typename Group::Type value() const { return value_; }
  void set_value(typename Group::Type value) { value_ = value; }
  void add_to_value(typename Group::Type delta) {
    set_value(Group::Plus(value(), delta));
  }
  void subtract_from_value(typename Group::Type delta) {
    set_value(Group::Minus(value(), delta));
  }
  //
  template<typename Node>
  static typename Group::Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, typename Group::Type value);
  template<typename Node>
  static void AddToAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void AddToProperAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromProperAnc(Node* node, typename Group::Type delta);

 protected:
  WithAncValue() : value_() {}

 private:
  typename Group::Type value_;
};

template<typename Semigroup, typename Base>
class WithAncAggr : public Base {
 public:
  typedef Semigroup Semigroup_;
  typename Semigroup::Type singleton_aggr() const {
    return Semigroup::AggrFromValue(this->value());
  }
  //
  template<typename Node>
  static typename Semigroup::Type AggrAnc(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostAnc(int dir, Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeafmostAnc(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRootmostAnc(Node* node, const Predicate& predicate);
  template<typename Node>
  static typename Semigroup::Type AggrProperAnc(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostProperAnc(int dir,
                                    Node* node,
                                    const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeafmostProperAnc(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRootmostProperAnc(Node* node, const Predicate& predicate);

 protected:
  WithAncAggr() {}
};

template<typename Base>
class WithEvertBy : public Base {
 private:
  typedef typename Base::Group_ Group;
 public:
  template<typename Node>
  static void Evert(Node* node);
 protected:
  WithEvertBy() {}
};

template<typename Base>
class WithEvert : public Base {
 public:
  template<typename Node>
  static void Evert(Node* node);
 protected:
  WithEvert() {}
};
}  // namespace naive
#endif  // DTREE_NAIVE_TREE_H_
