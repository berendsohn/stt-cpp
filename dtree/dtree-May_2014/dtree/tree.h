// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_TREE_H_
#define DTREE_TREE_H_

#include "dtree/selector.h"

#ifdef NAIVE_DTREE
#include "dtree/naive/tree.h"
#else
#include "dtree/selfadjust/tree.h"
#endif
#endif  // DTREE_TREE_H_
