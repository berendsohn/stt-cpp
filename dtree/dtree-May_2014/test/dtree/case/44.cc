#define DTREE_CONFIG EndTree<WithEvert<WithEvert<WithAncAggr<FreeSemigroup, NullPredicate_, WithAncAggr<FreeSemigroup, NullPredicate_, WithStaticValue<FreeValue, WithAncAggr<dtree::Count<int>, dtree::Index_<int>, Begin > > > > > > >
#include "test/dtree/tree.cc"
