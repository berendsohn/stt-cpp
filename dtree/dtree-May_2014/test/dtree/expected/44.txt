<<<
  1: v5d1,000,2ba,000 -> 7
  2: vdb2,000,a69,000 -> 14
  3: v720,000,b4e,000 -> 9
  4: v266,000,fc2,000 -> 15
  5: v78d,000,c62,000
  6: v9e2,000,171,000 -> 4
  7: va22,000,296,000 -> 12
  8: v79f,000,790,000
  9: v74f,000,e41,000 -> 15
  10: v702,000,acc,000 -> 15
  11: vd24,000,f9e,000 -> 6
  12: v84c,000,e1c,000
  13: vf7b,000,125,000 -> 6
  14: ve19,000,065,000 -> 3
  15: v172,000,030,000 -> 7
>>>
<<<
 7
  \1: a1 v5d1,000,2ba,000 a000,000 a000,000 f1 f0

  @5: a1 v78d,000,c62,000 a000,000 a000,000 f1 f0

  @8: a1 v79f,000,790,000 a000,000 a000,000 f1 f1

 15
  \10: a1 v702,000,acc,000 a000,000 a000,000 f1 f0

 6
  \11: a1 vd24,000,f9e,000 a000,000 a000,000 f0 f0

  @12: a1 v84c,000,e1c,000 a000,000 a000,000 f1 f1
      /7: a4 va22,000,296,000 a000,000 a000,000 f0 f0
        \15: a3 v172,000,030,000 a000,000 a000,000 f0 f0
          \6: a2 v9e2,000,171,000 a000,000 a000,000 f1 f0
            \4: a1 v266,000,fc2,000 a000,000 a000,000 f1 f1
    \13: a5 vf7b,000,125,000 a000,000 a000,000 f0 f0

 15
    /2: a1 vdb2,000,a69,000 a000,000 a000,000 f1 f1
  \14: a4 ve19,000,065,000 a000,000 a000,000 f0 f1
      /9: a1 v74f,000,e41,000 a000,000 a000,000 f0 f0
    \3: a2 v720,000,b4e,000 a000,000 a000,000 f0 f1
>>>
